File: frontend/app/public/pyModules/MoreStats.py
from abc import ABC, abstractmethod

class MoreStats(ABC):
    def __init__(self, primeFormCaluclator):
        self.pfc = primeFormCaluclator

    #takes a set (list) in ascending order
    #returns a vector (list) talling the instances of each
    #of the 6 interval classes
    @abstractmethod
    def getIntervalVector(self, set):
        pass

    #takes a set (list) in ascending order
    #returns all transpositions of the set as a list of lists
    @abstractmethod
    def getAllTranspositions(self, set):
        pass

    #use the getAllTranpositions method on the set
    #check if the set contains all the members of more 
    # than one of the transposions
    #if so, return true. otherwise false.
    #it might be helpful to convert the lists to sets for
    #easy comparison, but be careful to not shadow the set()
    #controcter in this scope. call the input something other
    #than set.

    #depends on self.getAllTranspositions
    @abstractmethod
    def isTranspositionallySymmetrical(self, setInput):
        pass


    #takes a set, returns the set inverted (12 - each member)
    #and in normal order tranposed to zero (Somolon's prime form)
    @abstractmethod
    def getInversion(self, set):
        pass

    #takes a set returns a boolean whether the set is
    #invesionally semmetrical
    #a set is inversionally semmetrical if the normal form
    #(solomon's prime form) of the set equals the normal 
    # form of the inversion.

    #depends on getInversion and PrimeForm.getPrimeForm
    @abstractmethod
    def isInversionallySymmetrical(self, set):
        pass
    
    #takes a set, returns the 12-tone complement in normal
    #form (solomon's prime form)

    #depends on PrimeForm.getPrimeForm
    @abstractmethod
    def getComplement(self, set):
        pass

    

File: backend-java-project/src/main/java/war/MoreStats.java
package war;

import java.util.ArrayList;

/**
 * MoreStats is an interface for analyzing pitch class sets.
 * 
 * All methods work with ArrayList<Integer>.
 * - Inputs: ArrayList<Integer> (usually in ascending order).
 * - Outputs: ArrayList<Integer> or boolean, depending on the method.
 * 
 */
public interface MoreStats {
    //Be sure to initialize this somewhere in your implementation
    //prior to the method calls for access to the instance methods 
    //of PrimeFormCalculator. Do this either in the constuctor or
    //in a static block
    PrimeForm pfc = null;

    /**
     * Takes a set (ArrayList<Integer>) in ascending order.
     * Returns a vector (ArrayList<Integer>) tallying the instances
     * of each of the 6 interval classes.
     *
     * @param set an ArrayList<Integer> representing a pitch class set.
     * @return an ArrayList<Integer> containing the interval vector.
     */
    public ArrayList<Integer> getIntervalVector(ArrayList<Integer> set);

    /**
     * Takes a set (ArrayList<Integer>) in ascending order.
     * Returns all transpositions of the set as an ArrayList of ArrayLists.
     *
     * @param set an ArrayList<Integer> representing a pitch class set.
     * @return an ArrayList<ArrayList<Integer>> containing all transpositions.
     */
    public ArrayList<ArrayList<Integer>> getAllTranspositions(ArrayList<Integer> set);

    /**
     * Uses the getAllTranspositions method on the set.
     * Checks if the set contains all the members of more than one of the transpositions.
     * If so, return true. Otherwise, return false.
     *
     * Depends on getAllTranspositions.
     *
     * @param setInput an ArrayList<Integer> representing a pitch class set.
     * @return true if the set is transpositionally symmetrical, false otherwise.
     */
    public boolean isTranspositionallySymmetrical(ArrayList<Integer> setInput);

    /**
     * Takes a set (ArrayList<Integer>).
     * Returns the set inverted (12 - each member) and put into normal order,
     * transposed to zero (Solomon's prime form).
     *
     * @param set an ArrayList<Integer> representing a pitch class set.
     * @return an ArrayList<Integer> containing the inversion in prime form.
     */
    public ArrayList<Integer> getInversion(ArrayList<Integer> set);

    /**
     * Takes a set (ArrayList<Integer>).
     * Returns a boolean indicating whether the set is inversionally symmetrical.
     *
     * A set is inversionally symmetrical if the normal form
     * (Solomon's prime form) of the set equals the normal form of the inversion.
     *
     * Depends on getInversion and pfc.getPrimeForm.
     *
     * @param set an ArrayList<Integer> representing a pitch class set.
     * @return true if the set is inversionally symmetrical, false otherwise.
     */
    public boolean isInversionallySymmetrical(ArrayList<Integer> set);

    /**
     * Takes a set (ArrayList<Integer>).
     * Returns the 12-tone complement in normal form (Solomon's prime form).
     *
     * Depends on pfc.getPrimeForm.
     *
     * @param set an ArrayList<Integer> representing a pitch class set.
     * @return an ArrayList<Integer> containing the complement in prime form.
     */
    public ArrayList<Integer> getComplement(ArrayList<Integer> set);
}
